File: weather.js
// 기상청 단기예보 — 산 위치 기준 시간대별 날씨 (흑백, 오프라인 스냅샷 지원).
// 도메인 로직(격자변환·발표시각·병합·아이콘)은 전부 여기 모아 iOS 이식 시 1:1 참조.
// 실제 API 호출은 CORS 때문에 프록시(/api/weather) 경유 — 네이티브는 직접 호출.

// ── 위경도 → 기상청 LCC 격자 (dfs_xy_conv) ──
export function dfsXy(lat, lon) {
  const RE = 6371.00877, GRID = 5.0, SLAT1 = 30, SLAT2 = 60, OLON = 126, OLAT = 38, XO = 43, YO = 136;
  const D = Math.PI / 180, re = RE / GRID, s1 = SLAT1 * D, s2 = SLAT2 * D, ol = OLON * D, oa = OLAT * D;
  let sn = Math.tan(Math.PI * 0.25 + s2 * 0.5) / Math.tan(Math.PI * 0.25 + s1 * 0.5);
  sn = Math.log(Math.cos(s1) / Math.cos(s2)) / Math.log(sn);
  let sf = Math.tan(Math.PI * 0.25 + s1 * 0.5);
  sf = (Math.pow(sf, sn) * Math.cos(s1)) / sn;
  let ro = Math.tan(Math.PI * 0.25 + oa * 0.5);
  ro = (re * sf) / Math.pow(ro, sn);
  let ra = Math.tan(Math.PI * 0.25 + lat * D * 0.5);
  ra = (re * sf) / Math.pow(ra, sn);
  let theta = lon * D - ol;
  if (theta > Math.PI) theta -= 2 * Math.PI;
  if (theta < -Math.PI) theta += 2 * Math.PI;
  theta *= sn;
  return { nx: Math.floor(ra * Math.sin(theta) + XO + 0.5), ny: Math.floor(ro - ra * Math.cos(theta) + YO + 0.5) };
}

// ── KST 시각 · 발표시각 계산 ──
function kstNow() { return new Date(Date.now() + (new Date().getTimezoneOffset() * 60000) + 9 * 3600000); }
const ymd = (d) => `${d.getFullYear()}${String(d.getMonth() + 1).padStart(2, "0")}${String(d.getDate()).padStart(2, "0")}`;

// 초단기예보: 매시 30분 발표, 45분 뒤 제공
function ultraBase() {
  const t = new Date(kstNow().getTime() - 45 * 60000);
  return { base_date: ymd(t), base_time: `${String(t.getHours()).padStart(2, "0")}30` };
}
// 단기예보: 02/05/08/11/14/17/20/23시 발표, +10분 제공
function vilageBase() {
  const t = new Date(kstNow().getTime() - 15 * 60000);
  const slots = [2, 5, 8, 11, 14, 17, 20, 23];
  const cand = slots.filter((s) => s <= t.getHours());
  if (cand.length) return { base_date: ymd(t), base_time: `${String(Math.max(...cand)).padStart(2, "0")}00` };
  const y = new Date(t.getTime() - 86400000);
  return { base_date: ymd(y), base_time: "2300" };
}

async function callKma(op, nx, ny, base) {
  const q = new URLSearchParams({ op, nx, ny, base_date: base.base_date, base_time: base.base_time });
  const r = await fetch(`/api/weather?${q}`);
  if (!r.ok) throw new Error("weather " + r.status);
  const d = await r.json();
  const items = d?.response?.body?.items?.item;
  if (!items) throw new Error("no items");
  return items;
}

// items → { "YYYYMMDDHHMM": {cat:val} } (fcst) 또는 실황
function byHour(items, fcst = true) {
  const m = {};
  for (const i of items) {
    const k = fcst ? i.fcstDate + i.fcstTime : "now";
    (m[k] ||= {})[i.category] = fcst ? i.fcstValue : i.obsrValue;
  }
  return m;
}

// ── 산 위치 시간대별 예보 조회 (초단기예보 + 단기예보 병합) ──
// 2시간 간격(지금·+2h·+4h…)으로 count 개. 반환: { baseLabel, hours: [...] }
export async function fetchWeather(lat, lon, { stepHours = 2, count = 8, spanHours = 24 } = {}) {
  const { nx, ny } = dfsXy(lat, lon);
  const [uf, vf] = await Promise.allSettled([
    callKma("ufcst", nx, ny, ultraBase()),  // 초단기예보(6h, 낙뢰 포함)
    callKma("vfcst", nx, ny, vilageBase()),  // 단기예보(오늘, POP·강수량)
  ]);
  const U = uf.status === "fulfilled" ? byHour(uf.value) : {};
  const V = vf.status === "fulfilled" ? byHour(vf.value) : {};
  if (!Object.keys(U).length && !Object.keys(V).length) throw new Error("예보 없음");

  const now = kstNow();
  const start = new Date(now.getFullYear(), now.getMonth(), now.getDate(), now.getHours());
  const num = (v) => (v == null || v === "" ? null : Number(v));
  const hours = [];
  for (let h = 0; h < spanHours && hours.length < count; h += stepHours) {
    const t = new Date(start.getTime() + h * 3600000);
    const key = ymd(t) + String(t.getHours()).padStart(2, "0") + "00";
    const u = U[key], v = V[key];
    if (!u && !v) continue;
    hours.push({
      key, hh: t.getHours(),
      tmp: num(u?.T1H) ?? num(v?.TMP),
      sky: num(u?.SKY) ?? num(v?.SKY),
      pty: num(u?.PTY) ?? num(v?.PTY) ?? 0,
      lgt: num(u?.LGT) ?? 0,               // 낙뢰(초단기 6h 만)
      pop: num(v?.POP),                    // 강수확률(단기만)
      pcp: v?.PCP && v.PCP !== "강수없음" ? v.PCP : (u?.RN1 && u.RN1 !== "강수없음" && num(u.RN1) ? u.RN1 : null),
      sno: v?.SNO && v.SNO !== "적설없음" ? v.SNO : null,
      wsd: num(u?.WSD) ?? num(v?.WSD),
    });
  }
  return { baseLabel: `${now.getHours()}시 기준`, savedAt: now.toISOString(), hours };
}

// ── 날씨 → 상태(아이콘 키·라벨). 우선순위: 낙뢰 > 강수 > 강풍 > 하늘 ──
export function weatherState(h) {
  if (h.lgt > 0) return { icon: "thunder", label: "낙뢰" };
  if (h.pty === 3) return { icon: "snow", label: "눈" };
  if (h.pty === 2) return { icon: "sleet", label: "비/눈" };
  if (h.pty === 1) return { icon: "rain", label: "비" };
  if (h.pty === 4) return { icon: "rain", label: "소나기" };
  if (h.wsd != null && h.wsd >= 9) return { icon: "wind", label: "강풍" };
  if (h.sky === 4) return { icon: "overcast", label: "흐림" };
  if (h.sky === 3) return { icon: "cloud", label: "구름많음" };
  return { icon: "sun", label: "맑음" };
}

// ── 흑백 SVG 아이콘 (currentColor stroke, 24뷰박스) ──
const P = (d, o = "") => `<path d="${d}" ${o}/>`;
const ICONS = {
  sun: `<circle cx="12" cy="12" r="4.4"/>${["M12 2.5v2.5", "M12 19v2.5", "M2.5 12H5", "M19 12h2.5", "M5.2 5.2l1.8 1.8", "M17 17l1.8 1.8", "M18.8 5.2L17 7", "M7 17l-1.8 1.8"].map((d) => P(d)).join("")}`,
  cloud: P("M7.5 18h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 10.2 3.9 3.9 0 0 0 7.5 18z"),
  overcast: P("M7.5 18h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 10.2 3.9 3.9 0 0 0 7.5 18z") + P("M4.5 14a3 3 0 0 1 1.2-2.3", 'opacity=".6"'),
  rain: P("M7.5 15h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 7.2 3.9 3.9 0 0 0 7.5 15z") + ["M8.5 17.5l-1 2.5", "M12 17.5l-1 2.5", "M15.5 17.5l-1 2.5"].map((d) => P(d)).join(""),
  snow: P("M7.5 15h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 7.2 3.9 3.9 0 0 0 7.5 15z") + ["M9 18.5v2.5M7.8 19.2l2.4 1.1M10.2 19.2l-2.4 1.1", "M15 18.5v2.5M13.8 19.2l2.4 1.1M16.2 19.2l-2.4 1.1"].map((d) => P(d)).join(""),
  sleet: P("M7.5 15h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 7.2 3.9 3.9 0 0 0 7.5 15z") + P("M8.5 17.5l-1 2.5") + P("M15 18.5v2.5M13.8 19.2l2.4 1.1M16.2 19.2l-2.4 1.1"),
  thunder: P("M7.5 14h9a3.5 3.5 0 0 0 .4-6.98A5 5 0 0 0 7.2 6.2 3.9 3.9 0 0 0 7.5 14z") + P("M12 15l-2 3.5h2l-1.2 3 3.2-4h-2l1.2-2.5z", 'fill="currentColor" stroke="none"'),
  wind: ["M4 10h9a2.2 2.2 0 1 0-2.1-2.8", "M4 14h13a2.4 2.4 0 1 1-2.3 3.1", "M4 12h6"].map((d) => P(d)).join(""),
};
export function weatherIcon(iconKey, size = 22) {
  return `<svg class="wx-svg" viewBox="0 0 24 24" width="${size}" height="${size}" fill="none" stroke="currentColor" stroke-width="1.6" stroke-linecap="round" stroke-linejoin="round">${ICONS[iconKey] || ICONS.sun}</svg>`;
}

// ── 시간대별 스트립 렌더 ──
// data: fetchWeather 결과, opts.offline: 스냅샷(오프라인) 표시 여부
export function renderStrip(el, data, opts = {}) {
  if (!el) return;
  if (!data || !data.hours || !data.hours.length) {
    el.innerHTML = `<div class="wx-empty">날씨 정보를 불러올 수 없습니다.</div>`;
    return;
  }
  const now = new Date(Date.now() + 9 * 3600000);
  const nowHH = new Date(now.getTime() + now.getTimezoneOffset() * 60000).getHours();
  const chips = data.hours.map((h, i) => {
    const s = weatherState(h);
    const timeLabel = i === 0 ? "지금" : `${h.hh}시`;
    const precip = h.pty ? `<div class="wx-pop">${h.pop != null ? h.pop + "%" : ""}${h.pcp ? " · " + h.pcp : (h.sno ? " · " + h.sno : "")}</div>` : "";
    return `<div class="wx-chip${i === 0 ? " now" : ""}">
      <div class="wx-time">${timeLabel}</div>
      <div class="wx-ic" title="${s.label}">${weatherIcon(s.icon)}</div>
      <div class="wx-tmp">${h.tmp != null ? Math.round(h.tmp) + "°" : "–"}</div>
      ${precip}
    </div>`;
  }).join("");
  const note = opts.offline
    ? `<span class="wx-note">오프라인 · ${data.savedAt ? new Date(data.savedAt).getHours() + "시 저장" : "저장된 예보"}</span>`
    : `<span class="wx-note">${data.baseLabel || ""} · 가장 가까운 관측지 기준</span>`;
  el.innerHTML = `<div class="wx-strip">${chips}</div><div class="wx-foot">${note}</div>`;
}

File: ios/HiHeight/Weather.swift
import Foundation

// 기상청 단기예보 — weather.js 1:1 직역(격자변환·발표시각·병합·상태).
// 실제 호출은 맥 admin_server 의 /api/weather 프록시 경유(키 은닉, IOS.md §2 전략 동일).
struct WeatherHour: Identifiable {
    let key: String
    let hh: Int
    let tmp: Double?
    let sky: Int?
    let pty: Int
    let lgt: Double
    let pop: Int?
    let wsd: Double?
    let isNow: Bool
    var id: String { key }
}

struct WeatherState { let symbol: String; let label: String }

enum WeatherService {
    // 위경도 → 기상청 LCC 격자 (dfs_xy_conv)
    static func dfsXy(lat: Double, lon: Double) -> (nx: Int, ny: Int) {
        let RE = 6371.00877, GRID = 5.0, SLAT1 = 30.0, SLAT2 = 60.0
        let OLON = 126.0, OLAT = 38.0, XO = 43.0, YO = 136.0
        let D = Double.pi / 180
        let re = RE / GRID, s1 = SLAT1 * D, s2 = SLAT2 * D, ol = OLON * D, oa = OLAT * D
        var sn = tan(.pi * 0.25 + s2 * 0.5) / tan(.pi * 0.25 + s1 * 0.5)
        sn = log(cos(s1) / cos(s2)) / log(sn)
        var sf = tan(.pi * 0.25 + s1 * 0.5)
        sf = pow(sf, sn) * cos(s1) / sn
        var ro = tan(.pi * 0.25 + oa * 0.5)
        ro = re * sf / pow(ro, sn)
        var ra = tan(.pi * 0.25 + lat * D * 0.5)
        ra = re * sf / pow(ra, sn)
        var theta = lon * D - ol
        if theta > .pi { theta -= 2 * .pi }
        if theta < -.pi { theta += 2 * .pi }
        theta *= sn
        return (Int(floor(ra * sin(theta) + XO + 0.5)), Int(floor(ro - ra * cos(theta) + YO + 0.5)))
    }

    private static var kst: Calendar {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "Asia/Seoul")!
        return c
    }
    // 예보 기준 캡션 — 현재 KST 시각 (weather.js baseLabel 대응).
    static var baseLabel: String { "\(kst.component(.hour, from: Date()))시 기준" }
    private static func ymd(_ d: Date) -> String {
        let c = kst.dateComponents([.year, .month, .day], from: d)
        return String(format: "%04d%02d%02d", c.year!, c.month!, c.day!)
    }
    // 초단기예보: 매시 30분 발표, 45분 뒤 제공
    private static func ultraBase(_ now: Date) -> (String, String) {
        let t = now.addingTimeInterval(-45 * 60)
        return (ymd(t), String(format: "%02d30", kst.component(.hour, from: t)))
    }
    // 단기예보: 02/05/08/11/14/17/20/23시 발표, +15분 여유
    private static func vilageBase(_ now: Date) -> (String, String) {
        let t = now.addingTimeInterval(-15 * 60)
        let h = kst.component(.hour, from: t)
        let slots = [2, 5, 8, 11, 14, 17, 20, 23].filter { $0 <= h }
        if let mx = slots.max() { return (ymd(t), String(format: "%02d00", mx)) }
        return (ymd(t.addingTimeInterval(-86400)), "2300")
    }

    private static func callKma(_ op: String, _ nx: Int, _ ny: Int, _ base: (String, String)) async -> [[String: Any]] {
        guard var comps = URLComponents(string: "\(Config.proxyBase)/api/weather") else { return [] }
        comps.queryItems = [
            .init(name: "op", value: op), .init(name: "nx", value: "\(nx)"), .init(name: "ny", value: "\(ny)"),
            .init(name: "base_date", value: base.0), .init(name: "base_time", value: base.1),
        ]
        guard let url = comps.url,
              let (data, _) = try? await URLSession.shared.data(from: url),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let body = (obj["response"] as? [String: Any])?["body"] as? [String: Any],
              let items = (body["items"] as? [String: Any])?["item"] as? [[String: Any]]
        else { return [] }
        return items
    }

    // items → { "YYYYMMDDHHMM": {category: value} }
    private static func byHour(_ items: [[String: Any]]) -> [String: [String: String]] {
        var m: [String: [String: String]] = [:]
        for i in items {
            guard let fd = i["fcstDate"] as? String, let ft = i["fcstTime"] as? String,
                  let cat = i["category"] as? String else { continue }
            m[fd + ft, default: [:]][cat] = "\(i["fcstValue"] ?? "")"
        }
        return m
    }

    // 미리 받아두기 — 결과는 버리고 URLCache 만 채운다(프록시 Cache-Control max-age=1800).
    // 탐험 진입 시점에 처음 요청하면 캐시 미스가 2초 넘게 걸려 날씨 칸이 비어 보인다.
    // 앱 시작·코스 캐러셀 진입처럼 **탐험보다 이른 시점**에 걸어두면 도착할 때 이미 캐시에 있다.
    static func prefetch(lat: Double, lon: Double) {
        Task { _ = await fetch(lat: lat, lon: lon) }
    }

    // 산 위치 2시간 간격 count 개 예보 (초단기예보 + 단기예보 병합)
    static func fetch(lat: Double, lon: Double, count: Int = 8, stepHours: Int = 2) async -> [WeatherHour] {
        let (nx, ny) = dfsXy(lat: lat, lon: lon)
        let now = Date()
        async let u = callKma("ufcst", nx, ny, ultraBase(now))
        async let v = callKma("vfcst", nx, ny, vilageBase(now))
        let U = byHour(await u), V = byHour(await v)
        if U.isEmpty && V.isEmpty { return [] }

        var comps = kst.dateComponents([.year, .month, .day, .hour], from: now)
        comps.minute = 0; comps.second = 0
        guard let start = kst.date(from: comps) else { return [] }
        func numI(_ s: String?) -> Int? { s.flatMap { Int($0) } }
        func numD(_ s: String?) -> Double? { s.flatMap { Double($0) } }

        var hours: [WeatherHour] = []
        var h = 0
        while h < 24 && hours.count < count {
            let t = start.addingTimeInterval(Double(h) * 3600)
            let hh = kst.component(.hour, from: t)
            let key = ymd(t) + String(format: "%02d00", hh)
            let uu = U[key], vv = V[key]
            if uu == nil && vv == nil { h += stepHours; continue }
            hours.append(WeatherHour(
                key: key, hh: hh,
                tmp: numD(uu?["T1H"]) ?? numD(vv?["TMP"]),
                sky: numI(uu?["SKY"]) ?? numI(vv?["SKY"]),
                pty: numI(uu?["PTY"]) ?? numI(vv?["PTY"]) ?? 0,
                lgt: numD(uu?["LGT"]) ?? 0,
                pop: numI(vv?["POP"]),
                wsd: numD(uu?["WSD"]) ?? numD(vv?["WSD"]),
                isNow: hours.isEmpty))
            h += stepHours
        }
        return hours
    }

    // 상태 우선순위: 낙뢰 > 강수 > 강풍 > 하늘 (SF Symbol 매핑)
    static func state(_ h: WeatherHour) -> WeatherState {
        if h.lgt > 0 { return .init(symbol: "cloud.bolt", label: "낙뢰") }
        switch h.pty {
        case 3: return .init(symbol: "cloud.snow", label: "눈")
        case 2: return .init(symbol: "cloud.sleet", label: "비/눈")
        case 1: return .init(symbol: "cloud.rain", label: "비")
        case 4: return .init(symbol: "cloud.heavyrain", label: "소나기")
        default: break
        }
        if let w = h.wsd, w >= 9 { return .init(symbol: "wind", label: "강풍") }
        switch h.sky {
        case 4: return .init(symbol: "cloud", label: "흐림")
        case 3: return .init(symbol: "cloud.sun", label: "구름많음")
        default: return .init(symbol: "sun.max", label: "맑음")
        }
    }
}
